File: src/onboarding.py
"""Short onboarding → personal MOS plan."""

from __future__ import annotations

PURPOSES = ("work", "study", "family", "business")

_COPY = {
    "pl": {
        "title": "Start: 3 krótkie pytania",
        "sub": "Ułożymy oficjalną ścieżkę do MOS. Wniosek i tak złożysz tylko na mos.cudzoziemcy.gov.pl.",
        "q1": "Jaki jest cel pobytu?",
        "q2": "Masz już numer PESEL?",
        "q3": "Do kiedy masz legalny pobyt?",
        "q3_hint": "Jeśli nie wiesz — pomiń. Datę możesz dodać później w zakładce Teraz.",
        "yes": "Tak",
        "no": "Nie / nie wiem",
        "skip": "Nie wiem — pomiń",
        "next": "Dalej",
        "start": "Pokaż mój następny krok",
        "restart": "Zmień odpowiedzi",
        "purpose": {
            "work": "Praca",
            "study": "Studia",
            "family": "Rodzina",
            "business": "Biznes",
        },
        "summary_pesel_yes": "Masz PESEL — następny krok: Profil Zaufany / login.gov.pl (gov.pl).",
        "summary_pesel_no": "Najpierw PESEL (lub meldunek). Potem Profil Zaufany i oficjalny MOS.",
        "summary_due": "Przypomnimy przed końcem legalnego pobytu ({due}).",
        "summary_purpose": "Cel: {purpose}. Pokażemy typowe załączniki do MOS.",
    },
    "ru": {
        "title": "Старт: 3 коротких вопроса",
        "sub": "Составим официальный путь к MOS. Wniosek всё равно подаёшь только на mos.cudzoziemcy.gov.pl.",
        "q1": "Какая цель pobytu?",
        "q2": "У тебя уже есть PESEL?",
        "q3": "До какой даты легальный pobyt?",
        "q3_hint": "Если не знаешь — пропусти. Дату можно добавить позже во вкладке Teraz.",
        "yes": "Да",
        "no": "Нет / не знаю",
        "skip": "Не знаю — пропустить",
        "next": "Дальше",
        "start": "Показать мой следующий шаг",
        "restart": "Изменить ответы",
        "purpose": {
            "work": "Работа",
            "study": "Учёба",
            "family": "Семья",
            "business": "Бизнес",
        },
        "summary_pesel_yes": "PESEL есть — следующий шаг: Profil Zaufany / login.gov.pl (gov.pl).",
        "summary_pesel_no": "Сначала PESEL (или meldunek). Потом Profil Zaufany и официальный MOS.",
        "summary_due": "Напомним до конца легального pobytu ({due}).",
        "summary_purpose": "Цель: {purpose}. Покажем типичные приложения для MOS.",
    },
    "en": {
        "title": "Start: 3 short questions",
        "sub": "We build your official path to MOS. You still file only at mos.cudzoziemcy.gov.pl.",
        "q1": "What is your purpose of stay?",
        "q2": "Do you already have a PESEL?",
        "q3": "Until when is your legal stay valid?",
        "q3_hint": "If you do not know — skip. You can add the date later in the Now tab.",
        "yes": "Yes",
        "no": "No / not sure",
        "skip": "Not sure — skip",
        "next": "Next",
        "start": "Show my next step",
        "restart": "Change answers",
        "purpose": {
            "work": "Work",
            "study": "Studies",
            "family": "Family",
            "business": "Business",
        },
        "summary_pesel_yes": "You have PESEL — next: trusted profile / login.gov.pl (gov.pl).",
        "summary_pesel_no": "Start with PESEL (or meldunek). Then trusted profile and official MOS.",
        "summary_due": "We will remind you before legal stay ends ({due}).",
        "summary_purpose": "Purpose: {purpose}. We will show typical MOS attachments.",
    },
    "ua": {
        "title": "Старт: 3 короткі питання",
        "sub": "Складемо офіційний шлях до MOS. Wniosek усе одно подаєш лише на mos.cudzoziemcy.gov.pl.",
        "q1": "Яка мета pobytu?",
        "q2": "У тебе вже є PESEL?",
        "q3": "До якої дати легальний pobyt?",
        "q3_hint": "Якщо не знаєш — пропусти. Дату можна додати пізніше у вкладці Teraz.",
        "yes": "Так",
        "no": "Ні / не знаю",
        "skip": "Не знаю — пропустити",
        "next": "Далі",
        "start": "Показати мій наступний крок",
        "restart": "Змінити відповіді",
        "purpose": {
            "work": "Робота",
            "study": "Навчання",
            "family": "Сім'я",
            "business": "Бізнес",
        },
        "summary_pesel_yes": "PESEL є — наступний крок: Profil Zaufany / login.gov.pl (gov.pl).",
        "summary_pesel_no": "Спочатку PESEL (або meldunek). Потім Profil Zaufany і офіційний MOS.",
        "summary_due": "Нагадаємо до кінця легального pobytu ({due}).",
        "summary_purpose": "Мета: {purpose}. Покажемо типові додатки для MOS.",
    },
}


def copy_for(lang: str) -> dict:
    return _COPY[lang if lang in _COPY else "pl"]


def build_plan(
    purpose: str,
    has_pesel: bool,
    due_at: str | None = None,
    lang: str = "pl",
) -> dict:
    purpose = purpose if purpose in PURPOSES else "work"
    c = copy_for(lang)
    purpose_label = c["purpose"][purpose]
    lines = [
        c["summary_purpose"].format(purpose=purpose_label),
        c["summary_pesel_yes"] if has_pesel else c["summary_pesel_no"],
    ]
    if due_at:
        lines.append(c["summary_due"].format(due=due_at[:10]))
    first = "trusted_profile" if has_pesel else "pesel"
    return {
        "purpose": purpose,
        "has_pesel": has_pesel,
        "due_at": (due_at or "")[:10] or None,
        "first_step": first,
        "summary_lines": lines,
        "copy": c,
    }

File: landing/onboarding.js
(() => {
  const API = "";
  const PLAN_KEY = "wniosekpl_onboarding_plan";
  const PURPOSE_KEY = "wniosekpl_mos_purpose";
  const READY_KEY = "wniosekpl_mos_ready";
  const TOKEN_KEY = "wniosekpl_session_token";
  const USER_KEY = "wniosekpl_web_user_id";
  const LANG_KEY = "wniosekpl_lang";

  let state = {
    purpose: localStorage.getItem(PURPOSE_KEY) || "",
    has_pesel: null,
    due_at: "",
  };
  let ui = null;

  function lang() {
    return localStorage.getItem(LANG_KEY) || "pl";
  }
  function token() {
    return localStorage.getItem(TOKEN_KEY) || "";
  }
  function userId() {
    return Number(localStorage.getItem(USER_KEY) || 0) || undefined;
  }
  function headers() {
    const h = { "Content-Type": "application/json" };
    if (token()) h.Authorization = `Bearer ${token()}`;
    return h;
  }
  function loadPlan() {
    try {
      return JSON.parse(localStorage.getItem(PLAN_KEY) || "null");
    } catch (_) {
      return null;
    }
  }
  function savePlan(plan) {
    localStorage.setItem(PLAN_KEY, JSON.stringify(plan));
  }
  function loadReady() {
    try {
      return JSON.parse(localStorage.getItem(READY_KEY) || "{}");
    } catch (_) {
      return {};
    }
  }
  function saveReady(ready) {
    localStorage.setItem(READY_KEY, JSON.stringify(ready));
  }

  function showStep(n) {
    ["1", "2", "3"].forEach((i) => {
      const el = document.getElementById(`ob-step-${i}`);
      if (el) el.hidden = String(n) !== i;
    });
    const done = document.getElementById("ob-done");
    if (done) done.hidden = n !== "done";
  }

  function setWorkspaceVisible(on) {
    const workspace = document.getElementById("now-workspace");
    const ob = document.getElementById("onboarding");
    if (workspace) workspace.hidden = !on;
    if (!ob) return;
    if (on) {
      ob.classList.add("ob-compact");
      // slim bar: title + restart only
      ["ob-sub", "ob-step-1", "ob-step-2", "ob-step-3", "ob-summary"].forEach((id) => {
        const el = document.getElementById(id);
        if (el) el.hidden = true;
      });
      const done = document.getElementById("ob-done");
      if (done) done.hidden = false;
      const start = document.getElementById("ob-start");
      if (start) start.hidden = true;
      const kicker = document.getElementById("ob-kicker");
      if (kicker) kicker.textContent = "Plan";
    } else {
      ob.classList.remove("ob-compact");
      const start = document.getElementById("ob-start");
      if (start) start.hidden = false;
      const sub = document.getElementById("ob-sub");
      if (sub) sub.hidden = false;
    }
  }

  function applyCopy(copy) {
    ui = copy;
    const map = [
      ["ob-title", "title"],
      ["ob-sub", "sub"],
      ["ob-q1", "q1"],
      ["ob-q2", "q2"],
      ["ob-q3", "q3"],
      ["ob-q3-hint", "q3_hint"],
      ["ob-pesel-yes", "yes"],
      ["ob-pesel-no", "no"],
      ["ob-deadline-next", "next"],
      ["ob-deadline-skip", "skip"],
      ["ob-start", "start"],
      ["ob-restart", "restart"],
    ];
    map.forEach(([id, key]) => {
      const el = document.getElementById(id);
      if (el && copy[key]) el.textContent = copy[key];
    });
    const kicker = document.getElementById("ob-kicker");
    if (kicker) kicker.textContent = "1 / 3";
  }

  function renderPurposes() {
    const box = document.getElementById("ob-purposes");
    if (!box || !ui) return;
    box.innerHTML = "";
    Object.entries(ui.purpose || {}).forEach(([id, label]) => {
      const btn = document.createElement("button");
      btn.type = "button";
      btn.textContent = label;
      if (state.purpose === id) btn.classList.add("active");
      btn.addEventListener("click", () => {
        state.purpose = id;
        localStorage.setItem(PURPOSE_KEY, id);
        box.querySelectorAll("button").forEach((b) => b.classList.remove("active"));
        btn.classList.add("active");
        const kicker = document.getElementById("ob-kicker");
        if (kicker) kicker.textContent = "2 / 3";
        showStep(2);
      });
      box.appendChild(btn);
    });
  }

  function renderSummary(plan) {
    const ul = document.getElementById("ob-summary");
    if (!ul) return;
    ul.innerHTML = "";
    (plan.summary_lines || []).forEach((line) => {
      const li = document.createElement("li");
      li.textContent = line;
      ul.appendChild(li);
    });
  }

  async function finish() {
    const body = {
      user_id: userId(),
      purpose: state.purpose || "work",
      has_pesel: !!state.has_pesel,
      due_at: state.due_at || null,
      lang: lang(),
    };
    let plan;
    let steps = loadReady();
    try {
      const res = await fetch(`${API}/api/onboarding`, {
        method: "POST",
        headers: headers(),
        body: JSON.stringify(body),
      });
      const data = await res.json();
      if (!res.ok) throw new Error("onboarding");
      plan = data.plan;
      steps = { ...steps, ...(data.steps || {}) };
    } catch (_) {
      // offline fallback
      plan = {
        purpose: body.purpose,
        has_pesel: body.has_pesel,
        due_at: body.due_at,
        summary_lines: [
          (ui?.summary_purpose || "{purpose}").replace(
            "{purpose}",
            ui?.purpose?.[body.purpose] || body.purpose
          ),
          body.has_pesel ? ui?.summary_pesel_yes : ui?.summary_pesel_no,
        ].filter(Boolean),
        first_step: body.has_pesel ? "trusted_profile" : "pesel",
        copy: ui,
      };
      if (body.has_pesel) steps.pesel = true;
      if (body.due_at) steps.legal_stay = true;
    }
    saveReady(steps);
    localStorage.setItem(PURPOSE_KEY, plan.purpose);
    plan.completed = true;
    savePlan(plan);
    renderSummary(plan);
    const kicker = document.getElementById("ob-kicker");
    if (kicker) kicker.textContent = "✓";
    showStep("done");
  }

  function enterWorkspace() {
    const plan = loadPlan();
    if (plan) plan.entered = true;
    if (plan) savePlan(plan);
    setWorkspaceVisible(true);
    if (window.wniosekplLoadMos) window.wniosekplLoadMos().catch(() => {});
    if (window.wniosekplRefreshProfile) window.wniosekplRefreshProfile();
    if (window.wniosekplRefreshAccountLink) window.wniosekplRefreshAccountLink();
    document.getElementById("mos-next")?.scrollIntoView({ behavior: "smooth", block: "start" });
    // Soft nudge: one account across phone + browser.
    const banner = document.getElementById("tg-link-banner");
    if (banner && localStorage.getItem("wniosekpl_tg_banner_dismissed") !== "1") {
      banner.hidden = false;
    }
  }

  function resetOnboarding() {
    localStorage.removeItem(PLAN_KEY);
    state = { purpose: "", has_pesel: null, due_at: "" };
    setWorkspaceVisible(false);
    const kicker = document.getElementById("ob-kicker");
    if (kicker) kicker.textContent = "1 / 3";
    showStep(1);
    renderPurposes();
  }

  function wire() {
    document.getElementById("ob-pesel-yes")?.addEventListener("click", () => {
      state.has_pesel = true;
      const kicker = document.getElementById("ob-kicker");
      if (kicker) kicker.textContent = "3 / 3";
      showStep(3);
    });
    document.getElementById("ob-pesel-no")?.addEventListener("click", () => {
      state.has_pesel = false;
      const kicker = document.getElementById("ob-kicker");
      if (kicker) kicker.textContent = "3 / 3";
      showStep(3);
    });
    document.getElementById("ob-deadline-form")?.addEventListener("submit", async (e) => {
      e.preventDefault();
      state.due_at = document.getElementById("ob-deadline-date")?.value || "";
      await finish();
    });
    document.getElementById("ob-deadline-skip")?.addEventListener("click", async () => {
      state.due_at = "";
      await finish();
    });
    document.getElementById("ob-start")?.addEventListener("click", enterWorkspace);
    document.getElementById("ob-restart")?.addEventListener("click", resetOnboarding);
  }

  async function boot() {
    wire();
    try {
      const res = await fetch(`${API}/api/onboarding/copy?lang=${lang()}`);
      if (res.ok) {
        const data = await res.json();
        applyCopy(data.copy || {});
      }
    } catch (_) {
      applyCopy({
        title: "3 pytania — Twój plan",
        sub: "Od razu wiemy, od czego zacząć.",
        q1: "Jaki jest cel pobytu?",
        q2: "Masz już numer PESEL?",
        q3: "Do kiedy masz legalny pobyt?",
        q3_hint: "Jeśli nie wiesz — pomiń.",
        yes: "Tak",
        no: "Nie / nie wiem",
        skip: "Nie wiem — pomiń",
        next: "Dalej",
        start: "Zaczynam",
        restart: "Zmień odpowiedzi",
        purpose: { work: "Praca", study: "Studia", family: "Rodzina", business: "Biznes" },
      });
    }
    renderPurposes();

    const plan = loadPlan();
    if (plan?.completed) {
      renderSummary(plan);
      if (plan.entered) {
        setWorkspaceVisible(true);
        // keep a compact summary card
        showStep("done");
        const sub = document.getElementById("ob-sub");
        if (sub) sub.hidden = true;
      } else {
        setWorkspaceVisible(false);
        showStep("done");
      }
    } else {
      setWorkspaceVisible(false);
      showStep(1);
    }
  }

  window.wniosekplResetOnboarding = resetOnboarding;

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", boot);
  } else {
    boot();
  }
})();
